File: panel/app/services/frp_config.py
"""frpc 配置生成与隧道辅助服务。"""
from __future__ import annotations

import ipaddress
import json
from typing import Any

from ..core.config import get_settings
from ..core.errors import BizError, ConflictError, NotFoundError
from ..models import Domain, Node, Plan, Tunnel, User

settings = get_settings()

# 系统保留端口与常用占用端口，禁止分配给隧道
RESERVED_PORTS = set(range(1, 1024))
RESERVED_PORTS.update({7000, 8000, 9001, 5432, 3306, 6379, 27017})


def check_tunnel_limit(user: User, plan: Plan) -> None:
    """校验隧道数量额度。"""
    from ..models import Tunnel

    count = len(user.tunnels)
    if count >= plan.tunnel_limit:
        raise BizError(400, 2001, f"套餐最多可创建 {plan.tunnel_limit} 条隧道，请升级套餐或删除不再使用的隧道")


def check_domain_limit(user: User, plan: Plan) -> None:
    """校验免费域名数量额度。"""
    from ..models import Domain

    count = (
        user.id
        and __count_active_domains(user.id)
    )
    if count >= plan.domain_limit:
        raise BizError(400, 3003, f"套餐最多可申请 {plan.domain_limit} 个免费域名")


def __count_active_domains(user_id: int) -> int:
    from ..core.database import SessionLocal

    with SessionLocal() as db:
        from ..models import Domain

        return (
            db.query(Domain)
            .filter(Domain.user_id == user_id, Domain.status == "active")
            .count()
        )


def allocate_remote_port(db, node_id: int) -> int:
    """从节点端口池分配一个空闲远程端口。

    端口池：1024-65535 中的常用端口段（20000-39999 优先），
    排除保留端口与已被其他隧道占用的端口。
    """
    from ..models import Tunnel

    occupied = {
        port
        for (port,) in db.query(Tunnel.remote_port)
        .filter(Tunnel.node_id == node_id, Tunnel.remote_port.isnot(None))
        .all()
        if port
    }
    # 优先使用 20000-39999，避免撞上常见服务
    for start in (20000, 40000, 1024, 10000):
        for port in range(start, start + 1000):
            if port in RESERVED_PORTS or port in occupied:
                continue
            return port
    raise BizError(409, 3001, "节点远程端口已用尽，请联系管理员")


def check_remote_port_available(db, node_id: int, port: int, exclude_tunnel_id: int | None = None) -> bool:
    """校验远程端口是否可用。"""
    from ..models import Tunnel

    if port in RESERVED_PORTS:
        return False
    q = db.query(Tunnel).filter(Tunnel.node_id == node_id, Tunnel.remote_port == port)
    if exclude_tunnel_id:
        q = q.filter(Tunnel.id != exclude_tunnel_id)
    return q.first() is None


def check_subdomain_available(db, subdomain: str, exclude_domain_id: int | None = None) -> bool:
    """校验子域名是否被占用。"""
    q = db.query(Domain).filter(Domain.subdomain == subdomain, Domain.status == "active")
    if exclude_domain_id:
        q = q.filter(Domain.id != exclude_domain_id)
    return q.first() is None


def build_public_address(tunnel: Tunnel, node: Node | None = None) -> str:
    """构造公网访问地址。"""
    node = node or tunnel.node
    host = node.address if node else settings.panel_base_url
    if tunnel.type in ("http", "https"):
        domain = tunnel.custom_domain or tunnel.subdomain
        if domain:
            scheme = "https" if tunnel.type == "https" else "http"
            return f"{scheme}://{domain}"
        return f"{tunnel.type}://{host}:{tunnel.remote_port or ''}"
    if tunnel.remote_port:
        return f"{host}:{tunnel.remote_port}"
    return host


def generate_frpc_config(tunnel: Tunnel, user: User, node: Node) -> str:
    """生成单条隧道的 frpc.toml 配置。"""
    lines: list[str] = []
    lines.append("serverAddr = %r" % node.address)
    lines.append("serverPort = %d" % node.port)
    lines.append("")
    lines.append("auth.method = \"token\"")
    lines.append("auth.token = %r" % user_token_for_frpc(user))

    name = f"{tunnel.name}-{tunnel.id}"
    lines.append("")
    lines.append(f"[[proxies]]")
    lines.append(f"name = {name!r}")
    lines.append(f"type = {tunnel.type!r}")
    lines.append(f"localIP = {tunnel.local_ip!r}")
    lines.append(f"localPort = {tunnel.local_port}")
    if tunnel.remote_port:
        lines.append(f"remotePort = {tunnel.remote_port}")
    if tunnel.subdomain:
        lines.append(f"subdomain = {tunnel.subdomain!r}")
    if tunnel.custom_domain:
        lines.append(f"customDomains = [{tunnel.custom_domain!r}]")
    lines.append(f"transport.kcp = {'true' if tunnel.kcp else 'false'}")
    lines.append(f"transport.bandwidthLimit = {tunnel_bandwidth(user)}")
    lines.append(f"transport.bandwidthLimitMode = \"client\"")
    if tunnel.encryption:
        lines.append("transport.encryption.enable = true")
    else:
        lines.append("transport.encryption.enable = false")
    if tunnel.compression:
        lines.append("transport.compression.enable = true")
    if tunnel.secret_key:
        lines.append(f"secretKey = {tunnel.secret_key!r}")
    if tunnel.type == "loadbalance" and tunnel.load_balancers:
        lbs = json.loads(tunnel.load_balancers)
        lines.append("loadBalancer.group = \"weavenet-lb\"")
        lines.append(f"loadBalancer.groupKey = {str(tunnel.id)!r}")
        for i, lb in enumerate(lbs):
            lines.append("")
            lines.append(f"[[proxies]]")
            lines.append(f"name = {f'{tunnel.name}-lb{i}-{tunnel.id}'!r}")
            lines.append("type = \"tcp\"")
            lines.append(f"localIP = {lb.get('ip', '127.0.0.1')!r}")
            lines.append(f"localPort = {lb.get('port', 80)}")
            lines.append(f"remotePort = {tunnel.remote_port}")
            lines.append("loadBalancer.group = \"weavenet-lb\"")
            lines.append(f"loadBalancer.groupKey = {str(tunnel.id)!r}")

    return "\n".join(lines)


def tunnel_bandwidth(user: User) -> str:
    """按用户套餐返回限速值（Kbps）。"""
    plan = user.plan
    mbps = plan.speed_limit_mbps if plan else 8
    return f"{mbps * 1000}KB"


def user_token_for_frpc(user: User) -> str:
    """生成 frpc 连接鉴权 Token：基于用户 ID 与全局密钥的稳定签名。"""
    import hashlib

    raw = f"{user.id}:{user.password_hash}:{settings.secret_key}".encode()
    return "u_" + hashlib.sha256(raw).hexdigest()

File: panel-go/services/frp_config.go
package services

import (
	"encoding/json"
	"fmt"
	"strings"

	"gorm.io/gorm"

	"weavenet/panel/config"
	"weavenet/panel/database"
	"weavenet/panel/models"
	"weavenet/panel/utils"
)

// RESERVED_PORTS 保留端口。
var RESERVED_PORTS = func() map[int]bool {
	m := make(map[int]bool)
	for i := 1; i < 1024; i++ {
		m[i] = true
	}
	for _, p := range []int{7000, 8000, 9001, 5432, 3306, 6379, 27017} {
		m[p] = true
	}
	return m
}()

// CheckTunnelLimit 检查隧道数量额度。
func CheckTunnelLimit(db *gorm.DB, user *models.User, plan *models.Plan) error {
	var count int64
	db.Model(&models.Tunnel{}).Where("user_id = ?", user.ID).Count(&count)
	if count >= int64(plan.TunnelLimit) {
		return utils.ErrBiz(400, 2001, fmt.Sprintf("套餐最多可创建 %d 条隧道，请升级套餐", plan.TunnelLimit))
	}
	return nil
}

// CheckDomainLimit 检查域名额度。
func CheckDomainLimit(db *gorm.DB, user *models.User, plan *models.Plan) error {
	var count int64
	db.Model(&models.Domain{}).Where("user_id = ? AND status = ?", user.ID, "active").Count(&count)
	if count >= int64(plan.DomainLimit) {
		return utils.ErrBiz(400, 3003, "免费域名额度不足，请升级套餐")
	}
	return nil
}

// AllocateRemotePort 分配远程端口（四段扫描）。
func AllocateRemotePort(db *gorm.DB, nodeID uint) (int, error) {
	var occupied []int
	db.Model(&models.Tunnel{}).Where("node_id = ? AND remote_port IS NOT NULL", nodeID).Pluck("remote_port", &occupied)
	oc := make(map[int]bool)
	for _, p := range occupied {
		oc[p] = true
	}
	starts := []int{20000, 40000, 1024, 10000}
	for _, start := range starts {
		for i := 0; i < 1000; i++ {
			p := start + i
			if p > 65535 {
				break
			}
			if RESERVED_PORTS[p] || oc[p] {
				continue
			}
			return p, nil
		}
	}
	return 0, utils.ErrBiz(409, 3001, "节点远程端口已用尽")
}

// CheckRemotePortAvailable 检查远程端口是否可用。
func CheckRemotePortAvailable(db *gorm.DB, nodeID uint, port int, excludeID uint) bool {
	if RESERVED_PORTS[port] {
		return false
	}
	q := database.DB.Model(&models.Tunnel{}).Where("node_id = ? AND remote_port = ?", nodeID, port)
	if excludeID > 0 {
		q = q.Where("id != ?", excludeID)
	}
	var count int64
	q.Count(&count)
	return count == 0
}

// CheckSubdomainAvailable 检查子域名是否可用。
func CheckSubdomainAvailable(db *gorm.DB, subdomain string, excludeID uint) bool {
	q := db.Model(&models.Domain{}).Where("subdomain = ? AND status = ?", subdomain, "active")
	if excludeID > 0 {
		q = q.Where("id != ?", excludeID)
	}
	var count int64
	q.Count(&count)
	return count == 0
}

// BuildPublicAddress 构建公网访问地址。
func BuildPublicAddress(t *models.Tunnel, node *models.Node) string {
	if t.Type == "http" || t.Type == "https" {
		domain := ""
		if t.CustomDomain != nil && *t.CustomDomain != "" {
			domain = *t.CustomDomain
		} else if t.Subdomain != nil && *t.Subdomain != "" {
			domain = *t.Subdomain
		}
		if domain != "" {
			return t.Type + "://" + domain
		}
		return fmt.Sprintf("%s://%s:%d", t.Type, node.Address, t.RemotePort)
	}
	if t.RemotePort != nil {
		return fmt.Sprintf("%s:%d", node.Address, *t.RemotePort)
	}
	if node.Address != "" {
		return node.Address
	}
	return config.C.PanelBaseURL
}

// TunnelBandwidth 带宽字符串（KB）。
func TunnelBandwidth(mbps int) string {
	return fmt.Sprintf("%dKB", mbps*1000)
}

// GenerateFrpcConfig 生成 frpc.toml 配置文本。
func GenerateFrpcConfig(db *gorm.DB, t *models.Tunnel, user *models.User, node *models.Node) string {
	var sb strings.Builder
	token := utils.UserTokenForFrpc(user.ID, user.PasswordHash)
	sb.WriteString("serverAddr = \"" + node.Address + "\"\n")
	sb.WriteString(fmt.Sprintf("serverPort = %d\n", node.Port))
	sb.WriteString("auth.method = \"token\"\n")
	sb.WriteString(fmt.Sprintf("auth.token = \"%s\"\n", token))
	sb.WriteString("[[proxies]]\n")
	sb.WriteString(fmt.Sprintf("name = \"%s-%d\"\n", t.Name, t.ID))
	sb.WriteString(fmt.Sprintf("type = \"%s\"\n", t.Type))
	sb.WriteString(fmt.Sprintf("localIP = \"%s\"\n", t.LocalIP))
	sb.WriteString(fmt.Sprintf("localPort = %d\n", t.LocalPort))
	if t.RemotePort != nil {
		sb.WriteString(fmt.Sprintf("remotePort = %d\n", *t.RemotePort))
	}
	if t.Subdomain != nil && *t.Subdomain != "" {
		sb.WriteString(fmt.Sprintf("subdomain = \"%s\"\n", *t.Subdomain))
	}
	if t.CustomDomain != nil && *t.CustomDomain != "" {
		sb.WriteString(fmt.Sprintf("customDomains = [\"%s\"]\n", *t.CustomDomain))
	}
	sb.WriteString(fmt.Sprintf("transport.kcp = %v\n", t.KCP))
	if user.Plan != (models.Plan{}) {
		sb.WriteString(fmt.Sprintf("transport.bandwidthLimit = \"%s\"\n", TunnelBandwidth(user.Plan.SpeedLimitMbps)))
	}
	sb.WriteString("transport.bandwidthLimitMode = \"client\"\n")
	sb.WriteString(fmt.Sprintf("transport.encryption.enable = %v\n", t.Encryption))
	if t.Compression {
		sb.WriteString("transport.compression.enable = true\n")
	}
	if t.SecretKey != nil && *t.SecretKey != "" {
		sb.WriteString(fmt.Sprintf("secretKey = \"%s\"\n", *t.SecretKey))
	}
	if t.Type == "loadbalance" {
		var lbs []struct {
			IP   string `json:"ip"`
			Port int    `json:"port"`
		}
		if json.Unmarshal([]byte(t.LoadBalancers), &lbs) == nil {
			for i, lb := range lbs {
				sb.WriteString("[[proxies]]\n")
				sb.WriteString(fmt.Sprintf("name = \"%s-lb%d-%d\"\n", t.Name, i+1, t.ID))
				sb.WriteString("type = \"tcp\"\n")
				sb.WriteString(fmt.Sprintf("localIP = \"%s\"\n", lb.IP))
				sb.WriteString(fmt.Sprintf("localPort = %d\n", lb.Port))
				if t.RemotePort != nil {
					sb.WriteString(fmt.Sprintf("remotePort = %d\n", *t.RemotePort))
				}
				sb.WriteString("loadBalancer.group = \"weavenet-lb\"\n")
				sb.WriteString(fmt.Sprintf("loadBalancer.groupKey = \"%d\"\n", t.ID))
			}
		}
	}
	return sb.String()
}

// ParseLB 解析负载均衡后端列表。
func ParseLB(s string) []map[string]any {
	var out []map[string]any
	if s == "" {
		return out
	}
	var raw []struct {
		IP   string `json:"ip"`
		Port int    `json:"port"`
	}
	if json.Unmarshal([]byte(s), &raw) != nil {
		return out
	}
	for _, r := range raw {
		out = append(out, map[string]any{"ip": r.IP, "port": r.Port})
	}
	return out
}
